Add ROSDEP_INSTALL_OPTS env

wkentaro opened the pull request:
I need this to use package that have bad rosdep keys in `UPSTREAM_WORKSPACE` by `-r` to `rosdep install`.

130s commented on industrial_ci/src/tests/source_tests.sh:
```diff
@@ -126,8 +126,11 @@ fi
 
 ici_time_start rosdep_install
 
+if [ "${ROSDEP_INSTALL_OPTS// }" == "" ]; then
```
This `if` clause only defines a variable. So can it be moved to [industrial_ci/src/env.sh](https://github.com/ros-industrial/industrial_ci/blob/83f6c9ba12fdeb7333a73a72d221f681d263640e/industrial_ci/src/env.sh)?
Cc @ipa-mdl

ipa-mdl commented:
> I need this to use package that have bad rosdep keys in UPSTREAM_WORKSPACE by -r to rosdep install.

So why do you need to overwrite `-q` and `--ignore-src` in this case?
If an upstream package is broken, it would be great if you report or even fix it!

130s commented:
> So why do you need to overwrite `-q` and `--ignore-src` in this case?

I can't speak for @wkentaro and just am personally trying to think of how useful it can be. 
When upstream binary is unavailable (even for temporarily) but you want to somehow proceed your CI jobs, this change might be useful (regardless those jobs can pass or not) to work around the situation.

Anyways, generally speaking isn't it a good idea to give users an option to pass options to external commands like `rosdep`?

wkentaro commented:
Sorry for late.

> The situation you're trying to address is:
Packages installed via rosdep is not what you want. Instead you want to build from source in your workspace.
Is this correct? Your statement bad rosdep keys is a bit unclear to me.

Sorry for unclear comment, I mean deprecated or unreleased dependencies for "bad rosdep keys".
In that case, I want to pass options `-r` or `--skip-keys=xxx` to `rosdep`.
(My lab project has default option of `-r` because there are some unreleased or deprecated packages: https://github.com/jsk-ros-pkg/jsk_travis#environmental-variables)


> So why do you need to overwrite -q and --ignore-src in this case?

In this case, it is not needed.
But when someone think overwrite `-q` or `--ignore-src` in the future, it will need to break the api or behavior if I don't set `-q --ignore-src` as the default now.

ipa-mdl commented:
> Sorry for unclear comment, I mean deprecated or unreleased dependencies for "bad rosdep keys".

Deprecated and unreleased packages should not lead to "bad rosdep keys".
(assuming that you provide the unreleased packages from source)

> But when someone wants overwrite -q or --ignore-src in the future, it will need to break the api or behavior if I don't set -q --ignore-src as the default now.

I can't imaging any use case for disabling `--ignore-src` for CI tests.
`-q` is superfluous anyway, but it does not hurt.

wkentaro commented:
> I can't imaging any use case for disabling --ignore-src for CI tests.

I also think this is a rare case, but what about below case?
- you want to use a package (package_A) in a github repository which includes some packages (package_A,B,C)
- there is no dependency among package_A,B,C
- you want to build only package_A

In this case, you need to disable `--ignore-src`, to use `package_B`, `package_C` installed by apt.

ipa-mdl commented:
@wkentaro wrote:
> In this case, you need to disable --ignore-src, to use package_B, package_C installed by apt.

`rosdep` would install them by apt, but cakin will still use the package from `.travis.rosinstall`.
(I have added a blacklist feature for rosdep AND catkin in #137)
For now it would be better to remove or disable the packages in a `BEFORE_SCRIPT`.

@130s wrote:
> Anyways, generally speaking isn't it a good idea to give users an option to pass options to external commands like rosdep?

IMHO rosdep doesn't have any additional, useful flags that make sense for CI.
If we introduce this option now, #137 does not work anymore as this option would underminde the new features.

wkentaro commented:
> rosdep would install them by apt, but cakin will still use the package from .travis.rosinstall.
(I have added a blacklist feature for rosdep AND catkin in #137)
For now it would be better to remove or disable the packages in a BEFORE_SCRIPT.

Maybe you misunderstand my needs, and I don't need to disable the `--ignore-src` option, but `-r`.
I just added `--ignore-src` to the env variable for future possible changes, because we can ignore being built packages by catkin --blacklist as you said.

ipa-mdl commented:
> Maybe you misunderstand my needs, and I don't need to disable the --ignore-src option, but -r.

I strongly encourage you to not use `-r`, but ` --skip-keys "package_X, package_Y, package_Z"` for CI tests.
`rosdep` can fail for various reasons and ignoring them leads to broken builds.

What do you think of introducing a `ROSDEP_SKIP_KEYS` variable?
I could support this in #137 as well :)

> I just added --ignore-src to the env variable for future possible changes, because we can ignore being built packages by catkin --blacklist as you said.

You can blacklist the build, but `rostest` will use the packages from source space (I guess).

wkentaro commented:
> What do you think of introducing a ROSDEP_SKIP_KEYS variable?
I could support this in #137 as well :)

I agree that.

ipa-mdl commented:
>> What do you think of introducing a ROSDEP_SKIP_KEYS variable?

> I agree that.

here we go: #184 
Does this fit your needs?

wkentaro commented:
Yeah, looks good.

On Fri, Jun 30, 2017 at 8:41 PM, Mathias Lüdtke <notifications@github.com>
wrote:

> What do you think of introducing a ROSDEP_SKIP_KEYS variable?
>
> I agree that.
>
> here we go: #184
> <https://github.com/ros-industrial/industrial_ci/pull/184>
> Does this fit your needs?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ros-industrial/industrial_ci/pull/180#issuecomment-312246055>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEHFk_h3welzAccYkONffbiAn_E13fkJks5sJN77gaJpZM4NvIFJ>
> .
>



-- 
和田 健太郎 / Kentaro Wada
http://wkentaro.com

130s commented:
>>     Anyways, generally speaking isn't it a good idea to give users an option to pass options to external commands like rosdep?
>
> IMHO rosdep doesn't have any additional, useful flags that make sense for CI.
If we introduce this option now, #137 does not work anymore as this option would undermine the new features.

I'm not sure about this yet because:
- While trying to understand #137, I'm stuck at https://github.com/ros-industrial/industrial_ci/issues/106#issuecomment-313243334 so can't comment on #137 yet.
- Not sure how much/little `rosdep` commands we want to expose to end users. I often use some of them, e.g. `-y Tell the package manager to default to y or fail when installing`.

That said, #184 itself looks good so I merged it.

ipa-mdl commented:
I will close this for now, skipping keys was implemented in #184

130s commented:
Yeah, fair enough. Besides, thanks for the discussion @wkentaro @ipa-mdl
